### PR TITLE
InhertedTypeJsonConverter support for constructors + bug fixes

### PR DESCRIPTION
**Pull request by SurajGupta**
InhertedTypeJsonConverter now supports types with constructors
Also, in cases where there is ambiguity about which child type is properly represented by the JSON, the old methodology would pick a child-type instead of throwing

### CODE REVIEW

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 1]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+        ///   - Get all child types of the specified type.
+        ///   - Filter to child types where every 1st level JSON property is either a public (accessor is public)
+        ///     property or a public field of the child type, using case-insensitive matching.  Child types passing
+        ///     this filter are called 'candidates'.
+        ///   - If there are no candidates, then throw.
+        ///   - For all candidates, ask the serializer to deserialize the JSON as the candidate type.  Catch and 
+        ///     ignore exceptions when attempting to deserialize.
+        ///   - If more than one candidate successfully deserializes, then throw.
+        /// - Using the serializer to deserialize enables the method to support types with constructors,
+        ///   which JSON.net does well out-of-the-box and which would be consumbersome to emulate.
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: spelling - cumbersome

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 2]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+        /// - This method does not consider > 1st level JSON properties to determine candidates.  In other words,
+        ///   it is not matching on the fields/properties of the child's fields/properties (e.g. the match is done 
+        ///   on Dog.Owner, not Dog.Owner.OwnersAddress).  The issue is that that kind of matching would require
+        ///   complex logic.  For example, Strings have properties such as Length which would need to be ignored
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: spelling - properties

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 3]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+        ///   on Dog.Owner, not Dog.Owner.OwnersAddress).  The issue is that that kind of matching would require
+        ///   complex logic.  For example, Strings have properties such as Length which would need to be ignored 
+        ///   when reflecting.  Similarly, all fields/properites of value-types would need to be ignored.  There 
+        ///   are likely other corner-cases.  To keep things simple, if the 1st level properties match the type's
+        ///   properties and fields, by name, we hand-off the problem to the serializer and let it throw if
+        ///   there is some incompatability deep in the field/property hierarchy.
+        /// - This method does not consider the type of objects containted within JSON arrays to determine
+        ///   candidates.  This is difficult because JSON arrays can contain a mix of types (just likes .net 
+        ///   objects) and every element would have to be deserialized and matched against the child's IEnumerable
+        ///   type and undoubtedly complexity would arrise from dealing with generics and the vast implementations
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: spelling - arise

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 4]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+        ///   of IEnumerable.  Like the bullet above, we simply hand-off the problem to the serializer.
+        /// - If properties or fields are removed from a child type after it has been serialized, then the JSON
+        ///   will not deserialize properly because that child type will no longer be a candidate.  If, however,
+        ///   properties or fields are added to the child type, then the child type will continue to be a
+        ///   candidate for the serialized JSON.
+        /// - If the user serializes private or internal fields/properties, then this method will not work because
+        ///   it only looks for public fields/properties.  We cannot bank on the JSON having been serialized by
+        ///   the same serialized passed to this method.  Even if we could, the serializer is so highly
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: serializer

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 5]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+        ///   of IEnumerable.  Like the bullet above, we simply hand-off the problem to the serializer.
+        /// - If properties or fields are removed from a child type after it has been serialized, then the JSON
+        ///   will not deserialize properly because that child type will no longer be a candidate.  If, however,
+        ///   properties or fields are added to the child type, then the child type will continue to be a
+        ///   candidate for the serialized JSON.
+        /// - If the user serializes private or internal fields/properties, then this method will not work because
+        ///   it only looks for public fields/properties.  We cannot bank on the JSON having been serialized by
+        ///   the same serialized passed to this method.  Even if we could, the serializer is so highly 
+        ///   configurable that it would be difficult to determine whether or which internal or private fields
+        ///   or properites are serialized.
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: properties

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 6]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+                if (jsonProperties.All(p => childTypePropertiesAndFields.Contains(p)))
+                {
+                    candidateChildTypes.Add(childType);
+                }
+            }
 
-            var target = testObjects.Select(to => new { TestObject = to, Keys = testKeys[to] })
-                .OrderBy(t => t.Keys.Count)
-                .FirstOrDefault(t => !sourceKeys.Except(t.Keys).Any()); // Any keys in source that are not in the target object?
+            var deserizliedObjects = new List<object>();
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: spelling - deserializedObjects

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 7]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
             {
                 throw new JsonSerializationException(
                     string.Format(CultureInfo.InvariantCulture, "Unable to deserialize to type {0}, value: {1}", objectType, jsonObject));
             }
 
-            serializer.Populate(jsonObject.CreateReader(), target.TestObject.Instance);
+            if (deserizliedObjects.Count > 1)
+            {
+                var matchingTypes =
+                    deserizliedObjects.Select(_ => _.GetType().Name)
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: Should this be FullName?

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 8]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+                        .Aggregate((running, current) => running + " | " + current);
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: Won't this do " | FirstType | SecondType | ... | LastType"? I'd prefer to not have additional characters at the beginning or the end.

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 9]
```diff
@@ -54,34 +53,103 @@ public override bool CanConvert(Type objectType)
...
+            if (reader.TokenType == JsonToken.Null)
+            {
+                return null;
+            }
 
-            var testObjects = childTypes.Select(Activator.CreateInstance)
-                .Select(instance => new { Instance = instance, SerializedInstance = Serialize(serializer, instance) })
-                .ToList();
+            var jsonObject = JObject.Load(reader);
+            var childTypes = GetChildTypes(objectType).ToList();
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: This is only iterated one time from what I can tell so I'd drop the .ToList()

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs [hunk 10]
```diff
@@ -91,65 +159,16 @@ public override void WriteJson(JsonWriter writer, object value, JsonSerializer s
...
-            {
-                jsonSerializer.Serialize(writer, instance);
-                var json = writer.ToString();
-                var parsed = JObject.Parse(json);
-                return parsed;
-            }
-        }
-
-        private static IEnumerable<KeyValuePair<string, JToken>> GetKeys(JObject jObject)
+        private static List<string> GetProperties(JToken node)
```
Comment (vrobinson) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: Return type should be IReadOnlyCollection<string>

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    {
+        /// <summary>
+        /// As of 7.0.1, Json.NET suggests using a static instance for "stateless" contract resolvers, for performance reasons.
+        /// <a href="http://www.newtonsoft.com/json/help/html/ContractResolver.htm"/>
+        /// <a href="http://www.newtonsoft.com/json/help/html/M_Newtonsoft_Json_Serialization_DefaultContractResolver__ctor_1.htm"/>
+        /// "Use the parameterless constructor and cache instances of the contract resolver within your application for optimal performance."
+        /// Also, <a href="https://stackoverflow.com/questions/33557737/does-json-net-cache-types-serialization-information"/>
+        /// </summary>
+        public static CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver Instance
+            => new CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver();
```
Comment (vrobinson) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: As much as I love C# 6 syntax, I avoid it in recipes since it gets source included in other projects and I don't know what they will be using to build. Can you go back to old syntax on non-test projects?

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+        /// <inheritdoc />
+        protected override JsonProperty CreatePropertyFromConstructorParameter(JsonProperty matchingMemberProperty, ParameterInfo parameterInfo)
+        {
+            if (matchingMemberProperty == null)
+            {
+                string message = parameterInfo == null
+                                     ? "All constructor parameters are required; found one that is not specified in json"
+                                     : $"This constructor parameter is required, but not specified in json: {parameterInfo.Name}";
```
Comment (vrobinson) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: string.Format(CultureInfo.Invariant, ...)

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+                throw new JsonSerializationException(message);
+            }
+
+            var property = base.CreatePropertyFromConstructorParameter(matchingMemberProperty, parameterInfo);
+
+            if ((property != null) && (matchingMemberProperty != null))
```
Comment (vrobinson) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: matchingMemberProperty != null is already checked above. Remove redundant check here.

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+            {
+                var required = matchingMemberProperty.Required;
+                if (required == Required.Default)
+                {
+                    if ((matchingMemberProperty.PropertyType != null) &&
+                        matchingMemberProperty.PropertyType.IsValueType &&
+                        (Nullable.GetUnderlyingType(matchingMemberProperty.PropertyType) == null))
```
Comment (vrobinson) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: Remove redundant parenthesis

File: Recipes/JsonConfiguration/JsonConfiguration.cs [hunk 15]
```diff
@@ -63,7 +62,7 @@ public static JsonSerializerSettings CompactSerializerSettings
                 {
                     Formatting = Formatting.None,
                     NullValueHandling = NullValueHandling.Ignore,
-                    ContractResolver = new CamelCasePropertyNamesContractResolver(),
+                    ContractResolver = CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.Instance,
```
Comment (vrobinson) [inline Recipes/JsonConfiguration/JsonConfiguration.cs]: Is this is a problem since we are ignoring nulls on the one hand and requiring them on the other?

File: Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs [hunk 16]
```diff
@@ -238,5 +455,140 @@ private class InheritedType3 : IBaseInterface
...
+
+            public int TailLength;
+        }
+
+        [Bindable(true)]
+        private class Diet
+        {
+        }
+
+        private class Atkins : Diet
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs]: Why have an intermediate type that adds nothing?

File: Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs [hunk 17]
```diff
@@ -151,6 +151,223 @@ public void Serializer_deserializes_InheritedTypes()
...
+            var inheritedType3 = JsonConvert.DeserializeObject<IBaseInterface>(json, JsonConfiguration.DefaultSerializerSettings) as InheritedType3;
+
+            Assert.That(inheritedType3, Is.Not.Null);
+            Assert.That(inheritedType3.Int32, Is.EqualTo(50));
+            Assert.That(inheritedType3.Float, Is.EqualTo(.2f));
+            Assert.That(inheritedType3.String, Is.Null);
+        }
+
+        [Test]
+        public void Serializer_throws_JsonSerializationException_when_property_has_been_removed_from_child_type()
```
Comment (vrobinson) [inline Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs]: I'm not sure what this is testing that is any different. Maybe title is just throwing me. Is it? ..._when_json_has_extra_properties_not_available_on_any_child_type

Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs]: good suggestion, changed to "Serializer_throws_JsonSerializationException_when_json_has_extra_properties_not_available_on_any_child_type"

File: Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs (hunk 16 above)
Comment (SurajGupta) [inline Recipes.Test/JsonConfiguration/JsonConfigurationTest.cs]: I think we've already discussed on the phone.  Atkins is a well-known diet, and so its not parameterized by anything.  Whereas LowCalorie requires the number of calories.  And I need this type to test for deserializing empty json string {}

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs (hunk 11 above)
Comment (SurajGupta) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: GREAT catch.  yes, will fix up.

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs (hunk 12 above)
Comment (SurajGupta) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: fixed

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs (hunk 13 above)
Comment (SurajGupta) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: good catch

File: Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs (hunk 14 above)
Comment (SurajGupta) [inline Recipes/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver/CamelCasePropertyNamesWithConstructorParametersRequiredContractResolver.cs]: will fix - although I disagree with this stylistic approach - visually, the parens help me identify the distinct components that are being &&'ed together.

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 1 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 2 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 3 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 4 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 5 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 9 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: done

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 6 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 7 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: I could go either way - name reads nicer, but FullName is the most complete information.  I'll change to FullName

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 8 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: no it won't - that's the beauty of Aggregate

File: Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs (hunk 10 above)
Comment (SurajGupta) [inline Recipes/InheritedTypeJsonConverter/InheritedTypeJsonConverter.cs]: fixed

File: Recipes/JsonConfiguration/JsonConfiguration.cs (hunk 15 above)
Comment (SurajGupta) [inline Recipes/JsonConfiguration/JsonConfiguration.cs]: we've discussed - going to implement a new settings called MinimalSerializerSettings
